fms/modules/positions.py

```diff
@@ -243,20 +243,26 @@ def adjusted_qk(
             if use_cache and past_kv_state is not None:
                 position_ids += past_kv_state[0].size(2)
 
-        q_ = q.float().reshape(*q.size()[:-1], -1, 2)  # B H L D/2 2
-        k_ = k.float().reshape(*k.size()[:-1], -1, 2)  # B H L D/2 2
+        q_ = q.float().view(*q.size()[:-1], -1, 2)  # B H L D/2 2
+        k_ = k.float().view(*k.size()[:-1], -1, 2)  # B H L D/2 2
 
         # the max start position should be based on the max first position of each sequence
         max_start_pos = torch.max(position_ids[:, 0])
         alpha = self.compute_freqs_cis(q.device, max_start_pos + seq_len)
-        freqs = self.cached_freqs[q.device.index][alpha][position_ids].unsqueeze(1)
+        freqs = self.cached_freqs[q.device.index][alpha][position_ids]
 
         freqs = freqs.float()  # 1 1 L D/2 2 2
         q_out = (
-            freqs[:, :, -q.size(2) :, :, :, :].mul(q_.unsqueeze(-2)).sum(5).flatten(3)
-        )
+            freqs[:, -q.size(1) :, None, :, :, :]
+            .mul(q_.unsqueeze(-2))
+            .sum(5)
+            .flatten(3)
+        ).type_as(q)
```

Comment:
nit: we might want to break this up into multiple lines for readability

Reply:
We could also use einsum to make it a single call, although I don't know what the performance would looklike exactly (as this is just doing a matrix-vector multiplication)

Reply:
I think we should switch this to einsum in a separate small PR after this is merged